### PR TITLE
[e2e] Refactor and Move node related methods to framework/node package[1]

### PR DESCRIPTION
**Pull request by jiatongw**
**What type of PR is this?**
/kind cleanup
/priority backlog

**What this PR does / why we need it**:
This is the following up PR of #78282.
- Move `util.go` node related functions to `framework/node` package
- Replace `ExpectNoError()` calls with `return fmt.Errorf()` in the framework, and let the running tests call `ExpectNoError()`
- Fix references of some methods to use the right package

Reference issue #77197, #76206, #77197 (parent issue #75601 )

**Special notes for your reviewer**:
`GetReadySchedulableNodesOrDie()` this method called nearly hundreds times in tests. So in this PR, I just move this method(and methods called inside) to `framework/node/resource.go` but not point references to `e2enode` to avoid a `XXL` size PR. This work will be done in another following PR after this one is approved.

**Does this PR introduce a user-facing change?**

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/framework/node/resource.go
Comment (jiatongw) [inline test/e2e/framework/node/resource.go]: Here called Failf() but I think it's ok to keep this. This method `isNodeUntainted() ` returns a `bool` and it's easy to be called as `if isNodeUntainted(&n) { do sth }` (see callers `GetMasterAndWorkerNodesOrDie()` and `GetReadySchedulableNodesOrDie()` ) . Return a `error` type will make to logic complex and it's not necessary to do so.
Thoughts?

File: test/e2e/framework/node/resource.go
Comment (andrewsykim) [inline test/e2e/framework/node/resource.go]: Add TODO to remove references in framework/utils.go

File: test/e2e/framework/node/resource.go
Comment (jiatongw) [inline test/e2e/framework/node/resource.go]: According to https://github.com/kubernetes/kubernetes/pull/77155#discussion_r301851674
Moving this method here to unblock that PR

File: test/e2e/framework/node/resource.go
Comment (jiatongw) [inline test/e2e/framework/node/resource.go]: Unexported because no external package reference.

File: test/e2e/framework/node/resource.go
Comment (jiatongw) [reply test/e2e/framework/node/resource.go]: Addressed

File: test/e2e/framework/node/resource.go
Comment (jiatongw) [reply test/e2e/framework/node/resource.go]: revert this to exported as per golint warning: `exported func PodNodePairs returns unexported type []node.podNode, which can be annoying to use`

File: test/e2e/cloud/nodes.go
Comment (andrewsykim) [inline test/e2e/cloud/nodes.go]: This should be renamed now from `GetReadyNodesIncludingTaintedOrDie` to `GetReadyNodesIncludingTainted`

**Comment by andrewsykim** [comment]
All the `*OrDie` functions that now return an error should be renamed to exclude the `OrDie` part.

**Comment by draveness** [comment]
cc/

**Review by draveness — COMMENTED** [review]
cc/ @draveness

File: test/e2e/autoscaling/cluster_size_autoscaling.go
Comment (alejandrox1) [inline test/e2e/autoscaling/cluster_size_autoscaling.go]: This is an interesting change. `framework.ExpectNoError()` is essentially a wrapper for `ExpectNoErrorWithOffset()` which uses a [`gomega.ExpectWithOffset()`](https://godoc.org/github.com/onsi/gomega#ExpectWithOffset)
> This is most useful in helper functions that make assertions. If you want Gomega's error message to refer to the calling line in the test (as opposed to the line in the helper function) set the first argument of `ExpectWithOffset` appropriately.

The comments in `ExpectNoErrorWithOffset()` give a better example of this:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1370-L1371

Originally `e2enode.GetReadyNodesIncludingTaintedOrDie` makes the call to `gomega.ExpectWithOffset(1, err).()` which i reckon means that the error won't be logged as coming from `e2enode.GetReadyNodesIncludingTaintedOrDie` but from `getPoolNodes` in this case.
If we now make `e2enode.GetReadyNodesIncludingTaintedOrDie` return the error and `getPoolNodes` return the error, then error will be logged not by `getPoolNodes` but by the method that calls `getPoolNodes`.

My thinking is that we should keep the call to `gomega.ExpectWithOffset` in `e2enode.GetReadyNodesIncludingTaintedOrDie` and log the error in `getPoolNodes` otherwise we are changing the behavior of the tests and possibly affecting people's workflow.
Is this appropriate behavior?

wdyt?

I also mentioned this in one of my PRs https://github.com/kubernetes/kubernetes/pull/79753#discussion_r301208958
/cc @andrewsykim @timothysc

Comment (jiatongw) [reply test/e2e/autoscaling/cluster_size_autoscaling.go]: need more input from people

Comment (timothysc) [reply test/e2e/autoscaling/cluster_size_autoscaling.go]: +1 to @alejandrox1 's case, I'd say switch it back

File: test/e2e/cloud/nodes.go
Comment (timothysc) [reply test/e2e/cloud/nodes.go]: Same comment applies here.

**Review by timothysc — APPROVED** [review]
/approve  
^ To unblock but please address comments and others can lgtm

**Comment by draveness** [comment]
This PR needs rebase, please take another look, thanks!

**Comment by jiatongw** [comment]
So here is a complex problem: 
The original case is:
`GetReadyNodesIncludingTaintedOrDie` calls `waitListSchedulableNodesOrDie`, and `waitListSchedulableNodesOrDie` calls `ExpectNoError()`.  

Now `GetReadyNodesIncludingTaintedOrDie` should be in `framework/node` package, but `ExpectNoError()` is in `framework/util.go`, which is `framework` package. So in `framework/node` package, I need to import `framework`. So there is an import circle here. 

So those functions who call  `waitListSchedulableNodesOrDie` need to be stay in `framework/util.go`. Or we will need to copy `ExpectNoError()` to `node` package. Wha's your thoughts?

@timothysc @andrewsykim @alejandrox1

**Comment by alejandrox1** [comment]
> So here is a complex problem:
> The original case is:
> `GetReadyNodesIncludingTaintedOrDie` calls `waitListSchedulableNodesOrDie`, and `waitListSchedulableNodesOrDie` calls `ExpectNoError()`.
> 
> Now `GetReadyNodesIncludingTaintedOrDie` should be in `framework/node` package, but `ExpectNoError()` is in `framework/util.go`, which is `framework` package. So in `framework/node` package, I need to import `framework`. So there is an import circle here.
> 
> So those functions who call `waitListSchedulableNodesOrDie` need to be stay in `framework/util.go`. Or we will need to copy `ExpectNoError()` to `node` package. What's your thoughts?
> 
> @timothysc @andrewsykim @alejandrox1

One possible way of going about this would be to replace `framework.ExpectNoError()` with the actual code it that it calls.
soFor example, `framework.ExpectNoError()` from [framework/util.go:1361](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1366)
```go
func ExpectNoError(err error, explain ...interface{}) {
	ExpectNoErrorWithOffset(1, err, explain...)
}
```

calls `framework.ExpectNoErrorWithOffset()` from [framework/util:1372](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1372)
```go
func ExpectNoErrorWithOffset(offset int, err error, explain ...interface{}) {
	if err != nil {
		e2elog.Logf("Unexpected error occurred: %v", err)
	}
	gomega.ExpectWithOffset(1+offset, err).NotTo(gomega.HaveOccurred(), explain...)
}
```

Thus, one solution would be to replace a `framework.ExpectNoError()` with:
```go
if err != nil {
    e2elog.Logf("Unexpected error occurred: %v", err)
}
gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
```

File: test/e2e/framework/node/resource.go
Comment (alejandrox1) [reply test/e2e/framework/node/resource.go]: Funny enough, this seems to be called only in one other place and that place is within the framework/uitl.go https://github.com/kubernetes/kubernetes/search?q=isNodeUntainted&unscoped_q=isNodeUntainted

I would opt for a bool and an error.
A call to `isNodeUntainted` will get more wordy as one would need to check for an error but i think that being explicit about what errors to ignore and which ones to act on is something the calling code should explicitly prescribe.
This also seems like a useful method in general and having it decide for you may not make for the best user experience.

File: test/e2e/autoscaling/cluster_size_autoscaling.go
Comment (alejandrox1) [reply test/e2e/autoscaling/cluster_size_autoscaling.go]: moved this convo onto top level comments https://github.com/kubernetes/kubernetes/pull/79879#issuecomment-511402267

**Comment by andrewsykim** [comment]
@jiatongw I might be missing something, but is this true?

> GetReadyNodesIncludingTaintedOrDie calls waitListSchedulableNodesOrDie, and waitListSchedulableNodesOrDie calls ExpectNoError().

I don't see the call to ExpectNoError here:

```
// waitListSchedulableNodes is a wrapper around listing nodes supporting retries.
func waitListSchedulableNodes(c clientset.Interface) (*v1.NodeList, error) {
	var nodes *v1.NodeList
	var err error
	if wait.PollImmediate(poll, singleCallTimeout, func() (bool, error) {
		nodes, err = c.CoreV1().Nodes().List(metav1.ListOptions{FieldSelector: fields.Set{
			"spec.unschedulable": "false",
		}.AsSelector().String()})
		if err != nil {
			if testutils.IsRetryableAPIError(err) {
				return false, nil
			}
			return false, err
		}
		return true, nil
	}) != nil {
		return nodes, err
	}
	return nodes, nil
}

// waitListSchedulableNodesOrDie is a wrapper around listing nodes supporting retries.
func waitListSchedulableNodesOrDie(c clientset.Interface) (*v1.NodeList, error) {
	nodes, err := waitListSchedulableNodes(c)
	if err != nil {
		return nil, fmt.Errorf("error: %s. Non-retryable failure or timed out while listing nodes for e2e cluster", err)
	}
	return nodes, nil
}
```

**Comment by alejandrox1** [comment]
> @jiatongw I might be missing something, but is this true?
> 
> > GetReadyNodesIncludingTaintedOrDie calls waitListSchedulableNodesOrDie, and waitListSchedulableNodesOrDie calls ExpectNoError().
> 
> I don't see the call to ExpectNoError here:


@jiatongw refactored the expectnorError out of those functions 
> Replace ExpectNoError() calls with return fmt.Errorf() in the framework, and let the running tests call ExpectNoError()

**Comment by jiatongw** [comment]
@alejandrox1 For the `ExpectNoError()` issue, I think we can first log the error then call `gomega.Expect().NotTo()` , since `getPoolNodes` will log error in the original behavior. 

For the `isNodeUntainted()` issue, if we return error, we will have risk of changing the callers original behavior? `GetReadySchedulableNodesOrDie()` calls both `waitListSchedulableNodesOrDie` and `isNodeUntainted`

**Comment by jiatongw** [comment]
And in `WaitForAllNodesSchedulable ` method, the use case is to ignore the error/fail in ` isNodeUntainted` https://github.com/kubernetes/kubernetes/blob/8a1c9e2b414a527bc146b4d3889355c0149ee5d8/test/e2e/framework/util.go#L1998

**Comment by alejandrox1** [comment]
> @alejandrox1 For the `ExpectNoError()` issue, I think we can first log the error then call `gomega.Expect().NotTo()` , since `getPoolNodes` will log error in the original behavior.

I think that sounds good and it looks like the behavior that a consumer of the library would be content with.

> For the isNodeUntainted() issue, if we return error, we will have risk of changing the callers original behavior? GetReadySchedulableNodesOrDie() calls both waitListSchedulableNodesOrDie and isNodeUntainted

For functions such as that one, I, as a consumer of the library would prefer a bool and and error. I would decide on my code whether to ignore an error or not by checking if it is/isn't of a certain type. So this is a more UX-type change that we can leave for later (if you agree with it). So i guess just leave that one as is and please create an issue so that we can take another look at it after this PR merges.

**Comment by jiatongw** [comment]
Sounds good to me! @alejandrox1

File: test/e2e/framework/node/wait.go
Comment (jiatongw) [inline test/e2e/framework/node/wait.go]: After removing `OrDie` suffix, there were two naming `waitListSchedulableNodes` methods. So I changed this one to `checkWaitListSchedulableNodes`

File: test/e2e/cloud/nodes.go
Comment (jiatongw) [inline test/e2e/cloud/nodes.go]: Here I didn't replace with `gomega.Expect(err).NotTo(gomega.HaveOccurred())`, since it is called directly in a ginkgo test suit.  Please comment if I should replace `ExpectNoError` with `gomega.Expect(err)`

File: test/e2e/cloud/nodes.go
Comment (jiatongw) [inline test/e2e/cloud/nodes.go]: Here I didn't replace with `gomega.Expect(err).NotTo(gomega.HaveOccurred())`, since it is called directly in a ginkgo test suit.  Please comment if I should replace `ExpectNoError` with `gomega.Expect(err)`

**Comment by jiatongw** [comment]
To make it easy to review, I have a second commit here. @alejandrox1 @andrewsykim Please take a look, thanks!

File: test/e2e/autoscaling/cluster_size_autoscaling.go
Comment (jiatongw) [inline test/e2e/autoscaling/cluster_size_autoscaling.go]: I got the error from prow `e2e-verify`:

`The above files need to use framework.ExpectNoError(err) instead of 
Expect(err).NotTo(HaveOccurred()) or gomega.Expect(err).NotTo(gomega.HaveOccurred())`
Thought that we can ignore this?

File: test/e2e/cloud/nodes.go
Comment (alejandrox1) [reply test/e2e/cloud/nodes.go]: What do you mean by "called directly in a ginkgo test suit"?

My thinking is that for all the methods you rewrote to return an error (instead of calling `framework.ExpectNoError()` within them) should call `gomega.Expect(err).NotTo(gomega.HaveOcurred())`.
Otherwise the `framework.ExpectNoError(err)` will apply an "offset" and will return (upon an error) the line of the calling method (i.e., func calling `e2enode.GetReadyNodesIncludingTainted`) instead of this method (`e2enode.GetReadyNodesIncludingTainted`).

File: test/e2e/autoscaling/cluster_size_autoscaling.go
Comment (alejandrox1) [reply test/e2e/autoscaling/cluster_size_autoscaling.go]: The reason for this seems to have been [hack/verify-test-code.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/verify-test-code.sh) which I didn't know it existed.
I don't know if there is a broader conversation about instances where we can ignore this as it seems rather arbitrary.

Under the premise that we keep `hack/verify-test-cdoe.sh` as is we can:
1. Create a wrapper in the framework for `gomega.Expect(err).NotTo(gomega.HaveOccurred())`, something like:
```go
func ExpectNoErrorToOccur(err error, explain ...interface{}) {
	ExpectNoErrorWithOffset(0, err, explain...)
}
```
2. Call `framework. ExpectNoErrorWithOffset(0, err)`

Comment (jiatongw) [reply test/e2e/autoscaling/cluster_size_autoscaling.go]: I'd like to use the second way and leave a TODO in the comment. I was thinking open a tracking issue to track all the PRs will a TODO after we moving and renaming.

File: test/e2e/cloud/nodes.go
Comment (jiatongw) [reply test/e2e/cloud/nodes.go]: Got it

File: test/e2e/autoscaling/cluster_size_autoscaling.go
Comment (alejandrox1) [reply test/e2e/autoscaling/cluster_size_autoscaling.go]: +1 to that!

**Comment by jiatongw** [comment]
@alejandrox1 pls take another review, thanks!

**Review by alejandrox1 — COMMENTED** [review]
Thank you for getting all this done 🎉 
Left a couple comments but looks good!
/lgtm

File: test/e2e/apps/daemon_restart.go
Comment (jiatongw) [reply test/e2e/apps/daemon_restart.go]: Sure, will do

Comment (jiatongw) [reply test/e2e/apps/daemon_restart.go]: Keeping `OrDie` for `GetReadySchedulableNodesOrDie` because we are still using `framework.GetReadySchedulableNodesOrDie()` in the tests, and change to `e2enode` will make this PR a XXL size.. So I just copy this method here and I plan to do in a following PR to fix the references of this method.

Comment (alejandrox1) [reply test/e2e/apps/daemon_restart.go]: ah sorry, you did mentioned that at the beginning of the PR.

**Pull request merged**
